src/ppm.rs

```diff
@@ -0,0 +1,473 @@
+    for i in 0..num_shares - 1 {
+        getrandom(&mut helper[i][PROOF_SHARE])?;
+
+        let mut prng = Prng::from_seed(&helper[i][PROOF_SHARE])?;
+        for j in 0..proof_len {
+            let r = prng.next().unwrap();
+            leader_proof_share[j] -= r;
+        }
+
+        for j in 0..SEED_LENGTH {
+            helper[i][JOINT_RAND_HINT][j] ^= joint_rand_seed[j];
+        }
+    }
+
+    for j in 0..SEED_LENGTH {
+        leader_joint_rand_seed_hint[j] ^= joint_rand_seed[j];
+    }
+
+    // Prepare the output messages.
+    let mut aggregator_id = 0;
+    let mut msgs = Vec::with_capacity(num_shares);
+    msgs.push(PrioUpload {
+        aggregator_id,
+        input_share: PrioShare::Leader(leader_input_share),
+        proof_share: PrioShare::Leader(leader_proof_share),
+        joint_rand_seed_hint: leader_joint_rand_seed_hint,
+        blind: leader_blind,
+    });
+
+    for row in helper.into_iter() {
+        aggregator_id += 1;
+        let mut col = row.into_iter();
+        msgs.push(PrioUpload {
+            aggregator_id,
+            input_share: PrioShare::Helper(col.next().unwrap(), input_len),
+            proof_share: PrioShare::Helper(col.next().unwrap(), proof_len),
+            joint_rand_seed_hint: col.next().unwrap(),
+            blind: col.next().unwrap(),
+        });
+    }
+
+    Ok(msgs)
+}
+
+/// The message sent by an aggregator to every other aggregator. This is the final message of the
+/// input-validation protocol.
+#[derive(Clone)]
+pub struct PrioVerifier<F: FieldElement> {
+    /// The aggregator's share of the verifier message.
+    pub verifier_share: Vec<F>,
+
+    /// The aggregator's share of the joint randomness, derived from its input share.
+    pub joint_rand_seed_share: Vec<u8>,
+}
+
+/// The initial state of an aggregate upon receiving the [`PrioUpload`] message from the client.
+pub struct PrioAggregatorState<F, V>
+where
+    F: FieldElement,
+    V: Value<F>,
+{
+    input_share: V,
+    verifier_share: Vec<F>,
+    joint_rand_seed_hint: Vec<u8>,
+}
+
+/// Run by each aggregator, this consumes the [`PrioUpload`] message sent from the client and
+/// produces the [`PrioVerifier`] message that will be broadcast to the other aggregators. It also
+/// returns the aggregator's state.
+///
+/// # Parameters
+///
+/// * `msg` is the message sent by the client. It contains, among other things, an input and proof
+/// share.
+/// * `param` is used to reconstruct the input share from the raw message.
+/// * `query_rand_seed` is used to derive the query randomness shared by all the aggregators.
+pub fn prio_verify_start<F, V>(
+    msg: PrioUpload<F>,
+    param: V::Param,
+    query_rand_seed: &[u8],
+) -> Result<(PrioAggregatorState<F, V>, PrioVerifier<F>), PpmError>
+where
+    F: FieldElement,
+    V: Value<F>,
+{
+    assert_safe();
+
+    let input_share_data: Vec<F> = Vec::try_from(msg.input_share)?;
+    let input_share = V::try_from((param, &input_share_data)).map_err(|err| {
```

Comment:
It's unfortunate to throw away the structure of the underlying error. I assume the problem here is that the various `Value::try_from` implementations could yield all sorts of different errors? We could provide a variant on `PpmError` like `Conversion(&'static str, #[source] Box<dyn Debug>)`, and then do 

```
let input_share = V::try_from(...).map_err(|err| {
    PpmError::Conversion("can't convert the thing to a thing!", Box::new(err))
})
```
Then other crates like [`eyre_report`](https://docs.rs/eyre/0.6.5/eyre/struct.Report.html) can still print good error chains.